### PR TITLE
Solver overhaul: feature integration, improved performance, better hints.

### PR DESCRIPTION
**Pull request by mcg1969**
NOT TO BE MERGED YET.

This is an initial PR, building on my existing mgrant-update-deps branch, that attempts to solve a variety of issues related to features. 

The fundamental problem with conda's current feature implementation is that it operates in a linear fashion. First, it guesses which features are likely to be installed; second, it sets up the dependency tree using those features as a guide; and third, it solves the resulting problem. Unfortunately, when a package has different `track_features` values across different versions, the initial feature guess is likely to be wrong.

The solution offered here treats features as much as possible *like a dependency*. Consider a dependency `python 2.7*`; this means that a package matching that spec must be present in the solution. Well, consider a feature `mkl`; this means that a package providing `mkl` as a `track_feature` must be present in the solution. In that simple sense, features are identical to dependencies. But of course, they can be different in the following ways:

- Packages with different names can supply the same `track_features` value. 
- Packages with a `feature` are not to be selected unless the corresponding `track_feature` is already installed, or is pulled in by a package explicitly requested by a specification. So, for instance, `numpy` should never resolve to an MKL version unless _another_ installed package has the `mkl` track feature.

To the extent that the behavior of features mirrors dependencies, I attempted to re-used existing dependency search and resolution logic as much as possible. This resulted in considerable code simplification in places. Conversely, where the behaviors differ, there is a fair amount of reworking. So, for instance, we now need to do _two_ optimization passes, not one: the first minimizes the number of installed features, and the second performs the standard version maximization.

It's passing all the `tests/test_resolve.py` tests now, and it will soon pass all of the others. Still, we need to play with it, and apply it to some of the challenges found in solver-related issues (link to them in this PR discussion, please). So _we shouldn't merge_ until we have done so.

### CODE REVIEW

**Comment by msarahan** [comment]
ping @stuarteberg @ukoethe @pelson - as knowledgeable community members who have dealt with features, what do you guys think here?  Please ping anyone else who might also be interested.

**Comment by stuarteberg** [comment]
I'm a little out of my depth here; I haven't used features much.  If they have time to look, @ukoethe and @jakirkham could offer more useful feedback than me.

**Comment by mcg1969** [comment]
Well, if you've installed Python on Windows lately, you've used features. That's one of the larger pain points we're trying to fix. But this is turning into a rather significant refactoring _and simplification_. And that's a good thing, because we need more people who understand what is going on in this solver code.

**Comment by pelson** [comment]
I've just been playing with features to see how well they work in the case where they are enabled for a period, and then a "track_feature" is subsequently removed. It seems the current behaviour leaves a lot to be desired: 

 * Create package "my_meta_package" which tracks "feature_a" (note that a dash in the feature name causes conda build to fail)
 * Create package "a", v1, which has a feature called "feature_a"
 * Create package "a", v2, which doesn't have any features

----

 * conda install my_meta_package
 * conda install a -> v1 being installed. :+1: 

----

 * conda remove my_meta_package
 * conda update --all -> v1 remains installed, and is not updated to v2 :-1: 
 * The only way to update "a" is to ```conda update a``` directly

------

Is this the kind of behaviour you are addressing @mcg1969? I haven't had the chance to see if this PR fixes this particular issue, but can do so if that would be useful.

**Comment by ukoethe** [comment]
Thanks for pinging me on this important discussion. I think the most important task at the moment is to specify formally the rules how feature resolution is supposed to behave, such that there is a unique definition of what's going to happen in every possible situation. This should include if and when featured packages shall be selected, but also when error messages like 'unsatisfiable installation request' or 'ambiguous installation request' occur in the presence of features and which hints are then desirable. Currently, we only have a bunch of poorly documented heuristics, and that's not good enough.

IMHO, @mcg1969's approach to treat features like dependencies is correct. In fact, that's what I do explicitly when I work around the current limitations by means of version tags. That is, I create version numbers like `foo-1.2.3.vc14` to indicate that the package 'foo' defines `track_features: - vc14` or `features: - vc14`. Consequently, the normal version resolution algorithm can be used to enforce a feature as a dependency. Specifically, dependents of 'foo' will simply write
```
requirements:
  run:
    - foo  >=1.2,*.vc14*
```
The improved feature resolution algorithm should behave semantically like this, but should add the `,*.vc14*` part automatically where appropriate. (See also https://github.com/conda/conda/issues/1959 for a proposal how explicit version tags can be added automatically using jinja2 templates.)

Regarding @pelson's example above: After `conda remove my_meta_package`, there are two admissible policies for the subsequent `conda update --all` (neithor of which is currently realized): It could either replace package 'a' with its featureless variant 'v2', or it could keep the installed 'v1', but repair it by re-installing its dependency 'my_meta_package'. The rules mentioned above should exactly specify what's going to happen. Out of my head, I'd prefer the second resolution.

It is also important that the rules distinguish two fundamentally different cases:
* Packages that never define a certain feature. For example, a pure Python package doesn't typically define a compiler feature.
* Packages where a particular feature is only missing in certain versions, but defined in others. For example, numpy 1.10 might be available in 'mkl' and 'openblas' variants, but when upgrading to numpy 1.11, only the 'mkl' variant is supplied (at least temporarily). 

In the second case, the rules should have the effect that version 1.11 is excluded from all version resolution as long as `track_features: - openblas` is active in the current environment. This is currently not happening, see https://github.com/conda/conda/issues/1589.

I suggest that the conda designers come up with a draft for the rules how features should behave that we can then discuss.

**Comment by mcg1969** [comment]
Great comments, @ukoethe. It turns out I've gone down a bit of a rabbit hole here with regards to this fix. This is _good news_, in the sense that I'm seeing a _lot_ of simplification in the code, which in turn is further testimony that treating features like dependencies is the right choice. Furthermore, moving ahead, when we consider additional functionality in conda like `namespace` and `provides` constructs, the proper behavior is going to be easier to reason about.

This where I am converging right now. Features have one key difference compared to dependencies: features should not be installed unless they _have_ to be. For instance, if you install `python`, you know you're going to be selecting either `vc9`, `vc10`, or `vc14`. If you install `mkl`, you know you'll be getting the `mkl` track feature because its a metapackage created for that specific purpose. But if I _don't_ install `mkl` (or `accelerate`) it shouldn't select any of the packages that have the `mkl` feature.

What this has led to is the following _three-stage_ optimization pass in the solver when in "install" mode. 

- First, minimize the number of installed features in the solution.
- Second, maximize the version numbers of all requested packages.
- Third, minimize the number of dependencies installed.

At first this sounds expensive, but in fact, thanks to some concomitant solver improvements I've implemented, it's working great. "Update" and "remove" modes have similar multi-pass programs, but they get to use the same multi-pass optimization infrastructure. We'll be able to implement some really good policies here.

So bear with me, you're going to see a few more major commits on this PR before it's ready to merge, but I think we're all going to benefit from it.

**Comment by ukoethe** [comment]
> If you install mkl, you know you'll be getting the `mkl` track feature because its a metapackage created for that specific purpose. But if I don't install `mkl` (or accelerate) it shouldn't select any of the packages that have the mkl feature.

That's exactly the kind of rules that I would like to see in writing so that people don't have to reverse-engineer them. Thus, your rule would be:
* If a package comes in several variants, the feature-free variant is preferred, unless overridden by an active 'track_features' in the current environment.

This is reasonable, but doesn't cover the case when a package has only featured variants. Will this result in an 'ambiguous feature resolution' error (with suitable hints to install a 'track_features' package first), or will package developers have the possibility to specify the default choice explicitly? If the second, would it be reasonable to allow other packages to override the 'feature-free is preferred' default as well?

Moreover, what are the implications of three-stage optimization in more difficult situations? For example, when the latest version has no feature-free variant, but an older version has, there is a conflict between the goals 'minimize features' and 'maximize version numbers'. Your explanation of three-stage optimization suggests that the older feature-free version is chosen, but is this always guaranteed to happen or even desirable?

**Comment by mcg1969** [comment]
> If a package comes in several variants, the feature-free variant is preferred, unless overridden by an active 'track_features' in the current environment.

No, this isn't the rule. The rule is to _minimize the number of features required to satisfy the specifications. So for instance:

> This is reasonable, but doesn't cover the case when a package has only featured variants.

It _does_ cover this case. If a package only has featured variants, then its feature becomes a _required_ dependency, and the solver will pull in a package with the appropriate `track_features` setting. Which one will it pull in? It's possible this could be ambiguous, if multiple packages offer the `track_features` setting. And certainly, we should document that fact. But you'll always get a valid solution.

> If the second, would it be reasonable to allow other packages to override the 'feature-free is preferred' default as well?

I like the idea of being able to specify a non-feature version of a package. For instance, if you have `mkl` engaged, you might want a non-`mkl` variant of a package. It would be very easy to accomplish this with the solver; we simply need to figure out how to spec it.

> Moreover, what are the implications of three-stage optimization in more difficult situations? For example, when the latest version has no feature-free variant, but an older version has, there is a conflict between the goals 'minimize features' and 'maximize version numbers'.

Earlier stages (e.g., feature count) takes priority over later stages (e.g, version maximization).

Keep in mind: the solver shouldn't necessarily be tasked with handling every possible crazy corner case, as long as it always results in a dependency-compliant install. Much of what we need to do here is to _document_ best practices for using `features` and `track_features`. For instance, we already recommend that `track_features` be embedded in empty metapackages. We violate our own recommendation with `python`, but we are actually working on fixing this in other ways in the future.

As long as the common and reasonable cases work exactly as we expect, and as long as we document those cases where unexpected behavior may be seen, I think we're fine.

**Comment by mcg1969** [comment]
I've changed the title because frankly features are but one of the improvements here. The new pull request passes all tests except for a couple in `test_cli.py`.

**Comment by mcg1969** [comment]
I think we're in good shape now, @kalefranz, if we want to build a test package.

**Comment by ukoethe** [comment]
> we already recommend that `track_features` be embedded in empty metapackages.

I don't understand why metapackages should be empty. Assume `conda install visual-studio=14.0` activates `track_features: - vc14`. This package could easily install the appropriate runtime DLLs as well. What's the advantage of not doing so? Similarly, `conda install openblas` will install the openblas libraries. Why shouldn't it define something like `track_features: - blas_open_0_2` at the same time?

**Comment by mcg1969** [comment]
Well, it may be that the recommendation was put in place to work around some of the strange effects features were having using the older solver model. With this newer model, there may be no good argument for separating a feature out into a metapackage.

**Comment by jakirkham** [comment]
> we already recommend that `track_features` be embedded in empty metapackages.

It became the recommendation. I also agree with @ukoethe that it should no longer be the case. This was certainly one issue with the original model of features.

> I think we're in good shape now, @kalefranz, if we want to build a test package.

I would also be interested in taking this for a spin to see if this is working reasonably.

File: conda/resolve.py [hunk 1]
```diff
@@ -143,481 +164,429 @@ def __ge__(self, other):
...
-                    else:
-                        found = True
+    @memoize
+    def package_triple(self, fn):
+        if not fn.endswith('.tar.bz2'):
+            return self.package_triple(fn + '.tar.bz2')
+        rec = self.index.get(fn, None)
+        if rec is not None:
+            return (rec['name'], rec['version'], rec['build'])
+        return fn[:-8].rsplit('-',2)[0]
```
Comment (groutr) [inline conda/resolve.py]: Returning one element here causes problem with this line in solve() (line 714)
```
name, version, build = self.package_triple(pkg)
```

Comment (mcg1969) [inline conda/resolve.py]: Interesting. Can you give me an example where this is actually occurring? I'm not saying we shouldn't fix it, but I wonder if it should be fixed upstream, and maybe place an `assert` here.

Comment (groutr) [inline conda/resolve.py]: Here's my traceback:
```
❯ python cli/main.py install python
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ......
Solving package specifications: (u'psutil', u'3.4.1', u'py27_0')
(u'openssl', u'1.0.2e', u'0')
(u'requests', u'2.9.1', u'py27_0')
(u'traitlets', u'4.1.0', u'py27_0')
(u'python-dateutil', u'2.4.2', u'py27_0')
(u'conda-env', u'2.4.5', u'py27_0')
(u'decorator', u'4.0.6', u'py27_0')
(u'path.py', u'8.1.2', u'py27_1')
(u'pycosat', u'0.6.1', u'py27_0')
(u'python.app', u'1.2', u'py27_4')
(u'python', u'2.7.11', u'0')
(u'joy', u'1.0', u'py27_0')
(u'ipython', u'4.0.3', u'py27_0')
(u'pickleshare', u'0.5', u'py27_0')
(u'ecdsa', u'0.11', u'py27_0')
(u'clyent', u'1.2.0', u'py27_0')
(u'networkx', u'1.10', u'py27_0')
(u'appnope', u'0.1.0', u'py27_0')
(u'wheel', u'0.26.0', u'py27_1')
(u'jinja2', u'2.8', u'py27_0')
(u'pexpect', u'3.3', u'py27_0')
(u'setuptools', u'19.4', u'py27_0')
apache-libcloud
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "cli/main.py", line 210, in <module>
    main()
  File "cli/main.py", line 173, in main
    args_func(args, p)
  File "cli/main.py", line 180, in args_func
    args.func(args, p)
  File "/Users/xxxxx/projects/conda/conda/cli/main_install.py", line 45, in execute
    install.install(args, parser, 'install')
  File "/Users/xxxxxx/projects/conda/conda/cli/install.py", line 333, in install
    update_deps=args.update_deps)
  File "/Users/xxxxxx/projects/conda/conda/plan.py", line 412, in install_actions
    update_deps=update_deps):
  File "/Users/xxxxxx/projects/conda/conda/resolve.py", line 715, in solve
    name, version, build = self.package_triple(pkg)
ValueError: too many values to unpack
```

Comment (mcg1969) [inline conda/resolve.py]: I'm having trouble reproducing this. What is the value of your `installed` variable?

Comment (mcg1969) [inline conda/resolve.py]: Oops, I got it. I need to drop the `[0]` at the end

**Comment by mcg1969** [comment]
Thanks to @groutr who identified a couple of issues here. Also, thanks to his encouragement, I've done some profiling and implemented a couple of performance improvements.

**Comment by mcg1969** [comment]
I'm getting excited. Thus sucker is so much faster now on difficult cases (e.g., with forced downgrades).

**Comment by ijstokes** [comment]
@mcg1969 please see the comments I just added to what I think is the issue that led to this PR work in #1967.  Short version: most of the time most people (whether they know it or not, or would admit it or not) don't want previously installed packages to have their dependency constraints considered.  But when they do, they really do, and we need to warn them those constraints are being violated in a way that lets them make an informed decision about how to proceed without scaring them off.  And we need the right set of configuration options (`.condarc` and CLI based) so the user can dictate their preference (loosely: either "strict" or "flexible" dependency checking).

@kalefranz this is a big one for you to think about.

**Comment by mcg1969** [comment]
@ijstokes, I responded in the #1967 thread. Neither PR causes dependencies to be fully updated. Even with `--update-deps` enabled (which is the default), the _minimum_ set of dependency changes are sought. In particular, if the current dependencies are fine, they don't change at all.

**Comment by mcg1969** [comment]
We'll be creating a conda package soon. We will not tag this for release until interested parties have had a chance to put it through its paces. Note that this address some other environment-breaking behaviors, like explicit specifications and `conda remove`.

**Comment by jakirkham** [comment]
> We'll be creating a conda package soon.

Could you please ping us when you do?

> We will not tag this for release until interested parties have had a chance to put it through its paces.

Great looking forward to it.

**Comment by mcg1969** [comment]
Oh yes, I will be posting install instructions to both PR's and to several issue threads.

**Comment by mcg1969** [comment]
Thanks to @groutr for building conda packages for this release.
```
conda install -c https://conda.anaconda.org/rgrout conda
```
Installers available 32/64-bit Windows & Linux, 64-bit Mac. Please try it out on your most challenging install & update recipes.

**Comment by jakirkham** [comment]
**Note:** I'll keep updating this as I try things so that we can have a singular list of what works and what does not.

So, I went ahead and gave it a shot. Here are the issues I am seeing the following issues.

1. OpenBLAS feature appears to be ignored. ( https://github.com/conda/conda/issues/1476 )
2. Features not respected by update all. ( https://github.com/conda/conda/issues/1334 )

It looks like it solves the following issues.

1. Grabs the right version in an ambiguous case. ( https://github.com/conda/conda/issues/1338 )

**Comment by ukoethe** [comment]
> most of the time most people don't want previously installed packages to have their dependency constraints considered.

This surprises me. I certainly expected conda to never invalidate any installed packages, at least when these packages were installed explicitly (as opposed to automatically installed dependenies, which might silently change). Please provide easy options to enforce this behavior (if it is not the default).

**Comment by mcg1969** [comment]
@ukoethe, it is my intention to enforce consistency by default, and to identify and fix every case where it is not. The consensus among those actively involved with conda development agree with this. I don't necessarily think Ian would disagree either, _as long as_ we do a proper job of showing the user where this enforcement is causing them to miss out on newer versions of packages.

**Comment by ijstokes** [comment]
@ukoethe : to be clear, right now conda does not enforce any previous package dependencies.  The only thing you are (currently) guaranteed to have as "satisfied dependencies" are the last conda install command you invoked.  Think of it as `autoconf + deps`: if you go and do `./configure && make && sudo make install` you have no guarantees that this won't have broken a stack of other things that used the previous version of the software that you have just overwritten by your latest `autoconf`-based install sequence.  Conda gives you that experience without compilation and with the addition of dependencies being automatically fetched and installed.

@mcg1969 and @groutr the main problem we're going to have with this is that everyone currently has "old-style" environments where dependencies were not preserved.  I just installed conda from the `rgrout` channel and now I can't install anything else, AFAICT, since I get about 2 dozen conflicts.

And this emphasizes my point for @ukoethe : these supposed conflicts that are being reported really aren't conflicts that will result in brokenness: I know that my package will work just fine if it is installed, and that other things will be **no more broken** than they already are due to already violated dependencies resulting from installing the latest package.  The analogy I used elsewhere is that of dead hyperlinks: the web works because we don't enforce referential integrity.  Likewise, an operating system requires all package dependencies to be satisfied at all times or important things can be critically broken, however conda package dependencies are often too strict and enforcing them will prevent a user from installing packages in ways that **typically* won't break other things in the conda environment.  We need to allow "broken links".

**Comment by ukoethe** [comment]
@mcg1969 @ijstokes 
>  We need to allow "broken links".

I'm not yet convinced. I'd agree with your point of view if calling a conflicting package resulted in an error message like "Package xyz is outdated. Please call 'conda update xyz' to get the newest version" (the analog of a "page not found" error in your web analogy). However, in practice "broken link" often means that there is a subtle binary incompatibility between C++ extension modules (i.e. shared libraries) which causes a mysterious crash without useful error message whatsoever. In my experience, debugging these crashes takes forever, and enforcing DLL consistency in the first place is much more efficient. Also note that a major user group of our software are biologists, who are able to use conda properly, but lack the skills to differentiate between a DLL incompatibility in the installation environment and a normal software bug.

At the moment, I'd like to be able to configure a 'strict' installation policy that would refuse to execute conda commands which might break dependencies of already installed packages. When an installation fails this way, I can easily work around the problem by creating a new environment. This is much cleaner than crossing fingers in the hope that things will *typically* not break.

In the long run, the ideal solution might be a web-accessible database (e.g. at anaconda.org) of *known* compatibility information. The database will be filled by running test suites with different combinations of package versions and applying diagnostic tools such as [ABI Compliance Checker](http://lvc.github.io/abi-compliance-checker/). 'conda install' would then first consult the database, and only resort to the current heuristic policies if the database lacked the necessary information for the present situation. I know that this sounds like science fiction, but in the long run empirically derived compatibility information is probably the only clean solution without being overly restrictive.

**Pull request merged**